ocfweb/stats/printing.py

```diff
@@ -44,3 +46,53 @@ def _semester_histogram():
     ax.set_title('Remaining balances this semester')
 
     return fig
+
+
+@periodic(3600)
+def _toner_changes():
+    return [
+        (printer, _toner_changes_for_printer(printer))
+        for printer in ('papercut', 'pagefault', 'logjam', 'deforestation')
+    ]
+
+
+def _toner_changes_for_printer(printer):
+    with stats.get_connection() as cursor:
+        cursor.execute('''
+            CREATE TEMPORARY TABLE ordered1
+                (PRIMARY KEY (position))
+                AS (
+                    SELECT * FROM (
+                        SELECT
+                            T.*,
+                            @rownum := @rownum + 1 AS position
+                            FROM (
+                                (
+                                    SELECT * FROM printer_toner_public
+                                    WHERE printer = %s
+                                    ORDER BY date
+                                ) AS T,
+                                (SELECT @rownum := 0) AS r
+                            )
+                    ) AS x
+                )
+        ''', (printer,))
+        cursor.execute('''
+            CREATE TEMPORARY TABLE ordered2
+                (PRIMARY KEY (position))
+                AS (SELECT * FROM ordered1)
+        ''')
+        cursor.execute('''
+            SELECT
+                B.date AS date,
+                A.value as pages_before,
+                B.value as pages_after
+                FROM
+                    ordered1 as A,
+                    ordered2 as B
+                WHERE
+                    B.position = A.position + 1 AND
+                    B.value > A.value AND
+                    A.value > 0
+        ''')
```

Comment:
This seems overly complicated, I think it would be better to just show a graph of the information and then let the person viewing it figure out when toner changes were made, since they sometimes aren't even full refills. It's also harder with a table to get a sense of time between rows, since they are just dates, which a graph would help with. Depends on what purpose the statistics are being shown for though I suppose.